keps/sig-node/4369-allow-special-characters-environment-variable/README.md

```diff
@@ -0,0 +1,319 @@
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Allow all printable ASCII except '=', to be set as environment variables.
+
+## Motivation
+
+Kubernetes should not restrict which environment variable names can be used, because it has no way of knowing what the application may need, and people can't always choose their own variable names, which may limit the adoption of Kubernetes.
+
+### Goals
+
+* Allow users to set printable ASCII other than '=' as environment variables.
+
+### Non-Goals
+
+## Proposal
+
+* Implements relaxed validation at the top level validation method when validating API create requests.
```

Comment:
Can you detail what the new validation will be?  Is it "anything but `'='` or is it something more precise?

Reply:
> The summary says `Allow all printable ASCII except '='`. Is that well-defined?

I wrote a demo to verify the ASCII characters that can be set as environment variables
```
package main

import (
	"fmt"
	"os"
)

func setEnvVars() {
	for i := 0; i < 128; i++ {
		envKey := "a" + string(i)
		if err := os.Setenv(envKey, "abc"); err != nil {
			fmt.Println(i)
			fmt.Printf("set character %s to env name failed \n", envKey)
		}
	}
}

func main() {
	setEnvVars()
}

```
it results is
```
➜  ~ ./envvar
0
set character a to env name failed
61
set character a= to env name failed
```
It looks like all ASCII characters can be set as environment variables except for `NUL` with serial number 0 and `=` with serial number 61.

I think we can add a part that disallows `NUL` characters with serial number 0 in ASCII and `=` characters with serial number 61 .... But I'd like to check if the CRI API restricts them first.

Reply:
According to the test results in https://github.com/kubernetes/enhancements/pull/4373#discussion_r1468812791, I think we can constrain the characters to characters 33-127 in ASCII. I will add some regex to KEP, and elaborate it in more detail.